torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/matrix_ops.py

```diff
@@ -304,3 +311,102 @@ def sharded_bmm(args, kwargs, pg):
     extra_check=sharded_bmm_check,
     customized_func=sharded_bmm,
 )
+
+def sharded_layer_norm_check(*args, **kwargs):
+    """
+    Perform extra checks for the ``nn.LayerNorm`` op.
+    Ensure the normalized shape is compatible with
+    the size of the sharded tensor.
+
+    Args: same as ``torch.nn.LayerNorm``.
+
+    Return: None
+    """
+    st = args[0]
+    normalized_shape = args[1]
+    if st.dim() < len(normalized_shape):
+        raise ValueError(
+            "normalized_shape dim must not be greater than "
+            "the dim of the sharded tensor."
+        )
+    for idx in range(-1, -len(normalized_shape) - 1, -1):
+        if normalized_shape[idx] != st.size(idx):
+            raise ValueError(
+                f"Given normalized_shape=[{normalized_shape[idx]}], expected input with shape "
+                f"[*, {normalized_shape[idx]}], but got input of size {list(st.size())}."
+            )
+
+
+def sharded_layer_norm(args, kwargs, pg):
+    """
+    Handles ``__torch_function__`` dispatch for the ``torch.nn.LayerNorm`` op.
+    We gather all shards from local shards and perform a global normalization.
+    We then scatter the result back to each rank.
+
+    Args: same as ``torch.nn.LayerNorm``.
+
+    Return:
+        local_tensor (Tensor): New local tensor to build the sharded tensor.
+        sharding_spec (:class:`torch.distributed._shard.sharding_spec.ShardingSpec`):
+            sharding spec of the new sharded tensor.
+        new_st_size (torch.Size): Size of the new sharded tensor.
+    """
+    st = args[0]
+    normalized_shape = args[1]
+    local_tensor = st.local_tensor()
+    current_rank = dist.get_rank(pg)  # type: ignore[attr-defined]
+    global_tensor = (
+        torch.empty(st.size(), device=st.local_tensor().device)
+        if current_rank == 0
+        else None
+    )
+    st.gather(dst=0, out=global_tensor)
```

Comment:
This seems pretty inefficient, can't we only do one allreduce to compute the `mean` and then given the `mean` we can compute the `standard deviation` as well and do rest of the operations just locally?

Reply:
I see. sure, let me change it like that.